Week1 [STEP 3] mustseers

mustseers opened the pull request:
@Kioding 


1. 고민되었던 점 (알고자 했던점): 

   * 함수를 자유자재로 사용하기+문법익숙해지기 : 이번 과제를 통해 함수를 자유자재로 사용하고 싶은 욕망(?)과 욕구(?)가 있었는데, 그런 부분에 있어 많은 연습이 되었던 것 같습니다. 함수 내의 매개변수를 외부에서 읽어 들이지 못하는 부분을 이해하고 해결하는데 거의 3일을 소요했었지만 (일과 병행하느라...), 지나고보니 이런 시간들을 통해 함수 문법과 사용법에 익숙해질 수 있었습니다. 결국은, 상수선언 통해 정말 쉽게 해결할 수 있었는데, 그렇게나 많은 시간을 보냈다니.. 이래서 코딩의 세계는 심오하면서도 재미난거 같습니다.
   ex) let iceLength = 9 // 아이스크림 길이 선언 --> 함수로 불러 들여 사용 가능  
   
   * 함수를 최대한 쪼개보기 (야곰님 피드백): 함수를 최대한 쪼개어 코드 짜는 연습을 해보았습니다. 이런 연습들이 결국은 코드를 짤 때 조금 더 쉽게 논리적으로 사고할 수 있도록 도와준거 같습니다. 피드백 겟!!
   
   * Step2에서의 피드백 최대한 반영: 1) 함수명은 "동사+명사" 2) 의미없는 이름으로 naming 하지 않기 3) 순서도도 논리에 맞는 기호 사용하기+문장으로 적기 



2. 해결이 되지 않은 점

   * 토핑을 기재하지 않았을 때, 자동으로 공백 주기 --> if 문으로 해결 가능할 것 같은데.. 일단 이번에는 Pass
   
   * 아이스크림을 문자 4자 이상으로 했을 때, 자동으로 막대기를 가운데로 옮겨주기 --> if 문으로 해결 가능할 거 같은데.. count 도 해야할 거 같고.. 일단 이번에는 pass
   
   * 세세하게 다 성공하고 다음으로 넘어가고 싶으나, 작은 성공을 통해 앞으로 한발한발 나가주는 것도 중요할 것 같습니다. 너무 Week1에서 세세한 부분에 매몰되어 있으면, 제 스스로 지칠거 같아요. 이제는 Week2로 넘어가고, 해결되지 않은 점은 모아 두었다가 Week2를 끝내는 시점에 다시 돌아와서 해결해 보는 것도 좋을 것 같습니다. 그 때는 실력이 더 늘어서 간단히 해결할 수 있지 않을까요? 


3. 조언을 얻고 싶은 부분 등 상세하게 작성
   
   * 지금 주고 계신 격려와 응원도 충분하다고 생각합니다. 특히 지난주에는 포기할까 생각도 많이 했는데..
   
   * 제가 놓치고 있는 부분이 있다면 뭐든지 피드백 부탁 드립니다. 모든 피드백 겟할 마음 가짐 되어 있사옵니다.

mustseers committed:
Added WK1_Step3.drawio

Commit:
1. Week1 Step 3 과제 제출
2. Step 3 Flowchart 추가

Commit:
Pull the remote rep and merge

Merge branch '3_mustseers' of https://github.com/mustseers/swift-starter-week1 into 3_mustseers

Commit:
1. Week1 Step 3 과제물(Swift/main)
2. Flowchart.png 추가

Commit:
1. Week1 Step3 Flowchart 추가

Kioding commented on CodeStarterCamp_Week1/Step2.swift:
```diff
@@ -0,0 +1,21 @@
+//
+//  Step2.swift
+//  CodeStarterCamp_Week1
+//
+//  Created by Ares Jinhwan Kim on 2022/02/22.
+//
+
+import Foundation
+
+func makeIce(iceHeight: Int) { // Icecream
+    for _ in 1...iceHeight {
+        print("***********")
+    }
+
+}
+
+func makeIcebar(icebarHeight: Int) { // Icebar
+    for _ in 1...icebarHeight {
+        print("    | |     ")
+    }
+}
```
step2 에서 했던 파일을 따로 만들어서 관리해주고 있네요 👏

mustseers replied:
칭찬 감사합니다.

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,42 @@
+    print ("토핑: \(iceTopping)")
+    print ("막대길이: \(iceBarLength)")
+    print ()
+}
```
- 함수에 대한 주석을 작성한다면 함수 위에 작성해주시면 좋을 것 같습니다 :)
- 이미 잘 알고 계실텐데, 함수명은 동사 + 명사의 형태로 지어주는 게 좋은데요!
그렇게 네이밍을 지어줌으로써 함수의 역할이 무엇인지 알려줄 수 있답니다.
아이스크림 정보를 알려준다 라는 네이밍으로 바꿔볼 수 있을 것 같아요!
- `icecreamInfo ()` 의 함수에서 소괄호를 붙여주면 컨벤션에 좋을 것 같습니다. 
다른 함수도 통일시키면 좋을 것 같아요ㅎㅎㅎ

mustseers replied:
1. 함수의 소괄호는 함수 이름과 붙여 줄 것 (수정 완료)
2. 함수 이름은 “동사+명사” 형태로 할 것 (수정 완료)

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,42 @@
+    print ("몸통: \(iceBody)")
+    print ("토핑: \(iceTopping)")
+    print ("막대길이: \(iceBarLength)")
+    print ()
```
print 를 여러줄을 작성할 때는 다른 방법이 있는데요.
[Multi-line strings
](https://www.hackingwithswift.com/sixty/1/3/multi-line-strings) 이 부분을 참고하여 한번 적용해볼까요!?

mustseers replied:
print 여러줄 할 때는 “multiline strings” 활용 --> 상수선언을 통해 multiline 작성 후, print로 불러오기 (수정완료)

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,42 @@
+let iceLength: Int = 5 // 숫자 입력
+let iceBody: String = "&*&" // 문자 3자 입력
+let iceTopping: String = " " // 문자 1자 입력, 토핑 불필요시 공백 한칸 유지 " "
+let iceBarLength: Int = 5 // 숫자 입력
```
필요한 부분을 상수로 선언하여서 사용하고 있군요 🙌
그렇다면 혹시 형태가 서로 다른 빼빼로를 4개를 만들고 싶다면 어떤 방식으로 구현해볼 수 있을까요?

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,42 @@
+    for _ in 1...iceBarLength {
+            print (" | |")
+    }
 }
```
고정된 bar 라면 함수내부에서 상수를 선언하고 print 한다면
어떤 걸 출력하는 지 가독성이 높아질 수 있습니다 :)

mustseers replied:
상수를 변수로 선언하고, 변수에 입력값 넣어주기 (코드 수정 완료)

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,42 @@
 
+icecreamInfo()
+makeIcecream()
+makeIceBar()
```
각각의 함수를 호출을 해주고 있는데
이 순서를 그대로 또 하나의 함수를 묶어주면 3개를 호출하지 않아도
1개의 함수만으로도 호출이 가능해집니다.

주문을 받는다의 함수로 묶어줄 수 있을 것 같아요 :)

mustseers replied:
코드 수정 완료

Commit:
1. 함수 주석은 함수 위에 작성 ( 수정 완료)
2. 함수의 소괄호는 함수 이름과 붙여 줄 것 (수정 완료)
3. 함수 이름은 “동사+명사” 형태로 할 것 (수정 완료)
4. print 여러줄 할 때는 “multiline strings” 활용
    --> 상수선언을 통해 multiline 작성 후, print로 불러오기
5. 고정된 값이라면, 상수선언을 통해 가독성 높이기(수정완료)
6. 서로 다른 빼빼로 4개 만들기
    —-> 상수(let)를 변수(var)로 선언하고, 변수에 입력값 넣어다른 빼빼로 완성

Commit:
fix: 피드백 업데이트

1. 함수 주석은 함수 위에 작성 (수정 완료)
2. 함수의 소괄호는 함수 이름과 붙여 줄 것 (수정 완료)
3. 함수 이름은 “동사+명사” 형태로 할 것 (수정 완료)
4. print 여러줄 할 때는 “multiline strings” 활용
   --> 상수선언을 통해 multiline 작성 후, print로 불러오기 (수정 완료)
5. 고정된 값이라면, 상수선언을 통해 가독성 높이기 (수정 완료)
6. 서로 다른 아이스크림 4개 만들기 (수정 완료)
   -—> 상수(let)를 변수(var)로 선언하고, 변수에 입력값 넣어주기
   --> 서로 다른 아이스크림 만들기

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,73 @@
+    makeIceBody()
+    makeIceBar()
+    print()
+    print()
```
`print()` 는 무슨 역할일까요?

mustseers replied:
나오는 결과의 가독성을 높이기 위해(?) 한줄 띄어쓰기 위해 "print()"를 추가한 것입니다. 
개인적으로 다닥다닥 붙어 있는 걸 싫어해서... 그냥 붙여써야 하겠죠?

Kioding replied:
`combineBebero()` 는 전체적인 흐름을 한 눈에 볼 수 있는 곳 같은데 
출력하는 함수에서 한줄 더 띄어서 print 하고 `combineBebero()`에서는 지우는 편이 더 가독성이 있을 것 같아요!

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,73 @@
+        막대길이: \(iceBarLength)
+        """
+    print(iceInfo)
+}
```
`var` 로 선언한 `iceInfo` 는 변한 적이 없는 것 같아요.
전역변수로 선언한 iceLength, iceBody, iceTopping, iceBarLength 를 `var` 로 선언해주어서 
내부에서 호출하는 값이 바뀌는 것일 뿐, `iceInfo`에 형태나 폼이 바뀌지 않았기에 let 을 사용하라고 조언하는 것 같네요.

만약 추후의 코드에서 아이스크림 가격도  넣고싶다면 그 때는 `var` 를 사용하는 거겠죠?
```swift
func showIcecreamInfo() {
    var iceInfo: String = """
        길이: \(iceLength)
        몸통: \(iceBody)
        토핑: \(iceTopping)
        막대길이: \(iceBarLength)
        """
    iceInfo = """
        길이: \(iceLength)
        몸통: \(iceBody)
        토핑: \(iceTopping)
        막대길이: \(iceBarLength)
        아이스크림가격: 
        """
    print(iceInfo)
}

```

mustseers replied:
네. 어느 정도 느낌적으로 이해는 되는데, 조금 더 경험이 쌓여야 완벽하게 이해가 될 것 같습니다.

Kioding replied:
벌써 잘하고 계시는걸요! 한번에 이해한다면,,, 그건 초보자가 아니니까요ㅎㅎㅎ

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,73 @@
+iceBody = "^!^"      // 문자 3자 입력
+iceTopping = "&"     // 문자 1자 입력, 토핑 불필요시 공백 한칸 유지 " "
+iceBarLength = 3
+combineIce()
```
전역변수를 계속 바꿔가면서 활용하고 계시는 것 같아요.
이전에도 드린 코멘트지만 빼빼로를 100개를 만들어야한다면 이와 같은 형태를 100개를 추가하는 게 좋을까요?
`combineIce()` 에서 필요한 정보를 매개변수로 받아 그 내부의 함수들에게 할당해주는 방법으로 바꾸는 건 어떨까요?

`지역변수와 전역변수의 차이` 와 `매겨변수 활용법` 에 대해 알아보고 
전역변수` 를 사용하지 않는 형태로 고쳐보면 좋을 것 같습니다 😆

mustseers replied:
넵

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,20 +9,73 @@
-makeIce(iceHeight: 8) // Icecream 길이 조정
-makeIcebar(icebarHeight: 4) // Icebar 길이 조정
+//makeIce(iceHeight: 8) // Icecream 길이 조정
+//makeIcebar(icebarHeight: 4) // Icebar 길이 조정
```
테스트 코드일까요? 그렇다면 삭제하는 게 좋습니다 :)

mustseers replied:
Step 2 내용인데, 아까워서 안 지우고 있었습니다....(유물같이..)

Kioding replied:
다음 week 에서는 `branch` 로 관리하는 법을 해보면 좋을 것 같아요!
이 부분은 할 때 다시 안내해드릴게요ㅎㅎㅎ

Commit:
feat: 누드빼빼로 추가 <타입>: <제목>

1. 누드빼빼로 추가
2. 예시 4개 모두 완료
3. 코드 최적화(?)

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,73 +9,96 @@
-    for _ in 1...iceLength {
-        formIceBody(iceBodyIn: iceBody, iceToppingIn: iceTopping)
+    
+    if beberoBody == "|0|" {
+        
+        
+        // 입력된 bebero의 길이에 따라 몸통을 그리는 함수 (Nude)
+        func formBeberoBodyNude(beberoBodyIn: String, beberoToppingIn: String) {
+            print("\(beberoToppingIn)", "| |", "\(beberoToppingIn)", separator:"")
+            print("\(beberoToppingIn)", "\(beberoBodyIn)", "\(beberoToppingIn)", separator:"")
+        }
+        
+        
+        //beberocream을 그리는 함수
+        func makeBeberoBody() {
+            for _ in 1...beberoLength/2 {
+                formBeberoBodyNude(beberoBodyIn: beberoBody, beberoToppingIn: beberoTopping)
+            }
+        }
+          
+        
+        //beberocream 그리기 (합치기?)
+        func combineBebero() {
+            showBeberoInfo()
+            print()
+            makeBeberoBody()
+            makeBeberoBar()
+            print()
+            print()
+        }
```
누드 빼빼로의 경우를 나누기 위한 `if-else` 인 것 같은데 비슷한 함수를 2번씩 쓰고 있어 보입니다 👀
- if-else가 필요한 경우는 해당함수에서 분기처리를 해주면 하단의 함수를 다 지워도 될 것 같아요!
`formBeberoBodyNude()` 와 `formBeberoBody()` 를 하나로 통일하고
`makeBeberoBody()` 도 내부에서 분기처리해주고
- 그러면 `combineBebero()` 도 하나면 되겠죠?!

<br>

현재 함수 선언 부분을 헷갈리시는 것 같은데 함수는 따로 선언을 해준다고 생각하고
필요할 때만 호출해서 사용한다고 생각해주시면 될 것 같아요! (물론 이 부분이 쉽지 않죠 😱 ㅎㅎㅎ)

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,73 +9,96 @@
+            print()
+        }
+        
+        return combineBebero()
```
return 은 현재 사용하고 있지 않아보이는 데 어떤 의도로 작성하셨을까요?!

Kioding commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -9,73 +9,96 @@
+    }
+        
+    
+    else {
+
+        
+        // 입력된 beberocream의 길이에 따라 몸통을 그리는 함수
+        func formBeberoBody(beberoBodyIn: String, beberoToppingIn: String) {
+            print("\(beberoToppingIn)", "\(beberoBodyIn)", "\(beberoToppingIn)", separator:"")
+        }
+
+        
+        //beberocream을 그리는 함수
+        func makeBeberoBody() {
+            for _ in 1...beberoLength {
+                formBeberoBody(beberoBodyIn: beberoBody, beberoToppingIn: beberoTopping)
+            }
+        }
+          
+        
+        //beberocream 그리기 (합치기?)
+        func combineBebero() {
+            showBeberoInfo()
+            print()
+            makeBeberoBody()
+            makeBeberoBar()
+            print()
+            print()
+        }
+        
+        return combineBebero()
     }
```
이 부분은 위 코멘트를 반영한다면 지워도 될 것 같아요!
이 부분을 고민해봅시다 🥳 
여기까지 하시면 다음 step으로 넘어가도 충분할 것 같아요!
